types/src/v0/impls/stake_table.rs

```diff
@@ -26,84 +37,207 @@ use hotshot_types::{
     },
     PeerConfig,
 };
+use indexmap::IndexMap;
 use itertools::Itertools;
 use thiserror::Error;
 
 use super::{
     traits::StateCatchup,
-    v0_3::{DAMembers, StakeTable, StakeTables},
+    v0_3::{DAMembers, Delegator, StakeTable, Validator},
     Header, L1Client, Leaf2, NodeState, PubKey, SeqTypes,
 };
 
 type Epoch = <SeqTypes as NodeType>::Epoch;
 
-impl StakeTables {
-    pub fn new(stake_table: StakeTable, da_members: DAMembers) -> Self {
-        Self {
-            stake_table,
-            da_members,
+/// Create the consensus and DA stake tables from L1 events
+///
+/// This is a pure function, to make it easily testable.
+///
+/// We expect have at most a few hundred EVM events in the
+/// PermissionedStakeTable contract over the liftetime of the contract so it
+/// should not significantly affect performance to fetch all events and
+/// perform the computation in this functions once per epoch.
+pub fn from_l1_events(
+    registered: Vec<(ValidatorRegistered, Log)>,
+    deregistered: Vec<(ValidatorExit, Log)>,
+    delegated: Vec<(Delegated, Log)>,
+    undelegated: Vec<(Undelegated, Log)>,
+    keys_update: Vec<(ConsensusKeysUpdated, Log)>,
+) -> anyhow::Result<IndexMap<Address, Validator<BLSPubKey>>> {
+    let events = StakeTableEvent::sort_events(
+        registered,
+        deregistered,
+        delegated,
+        undelegated,
+        keys_update,
+    )?;
+
+    let mut validators = IndexMap::new();
+    let mut bls_keys = HashSet::new();
+    let mut schnor_keys = HashSet::new();
+    for event in events.values() {
+        match event {
+            StakeTableEvent::Register(ValidatorRegistered {
+                account,
+                blsVk,
+                schnorrVk,
+                commission,
+            }) => {
+                // TODO(abdul): BLS and Schnorr signature keys verification
+                let stake_table_key = bls_alloy_to_jf2(blsVk.clone());
+                let state_ver_key = edward_bn254point_to_state_ver(schnorrVk.clone());
+                if bls_keys.contains(&stake_table_key) {
+                    tracing::warn!("bls key {stake_table_key:?} already used");
```

Comment:
@imabdulbasit Can re return an error from this function for anything that is a bug in the smart contract? This would mean the confirmation layer encounters a bug in the contract so it's IMO best to stop running so that we can investigate. If we return an Error here and then panic in the outer layer at least we will be alerted that something is wrong. I think this is less dangerous than continuing operation after we find a bug in the contract. I think it will also make it more likely we discover bugs in testing and it makes this function easier to test on bad inputs.

Reply:
why is this a bug? Contract supports multiple BLS keys? I thought we would just ignore those events

Reply:
Currently the contract doesn't allow a BLS key to be used twice. We can of course allow it but it doesn't cost much gas to check and mark as used, it's true that we currently don't do it for schnorr keys but I think we should also do it for schnorr keys. I just catches some potential configuration errors and I don't think we really gain anything from allowing it in the contract.